File: app/src/main/java/com/urban/mvvmshowcase/view/MainActivity.java
package com.urban.mvvmshowcase.view;

import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;

import com.urban.mvvmshowcase.R;
import com.urban.mvvmshowcase.databinding.ActivityMainBinding;
import com.urban.mvvmshowcase.model.repository.FakePersonRepository;
import com.urban.mvvmshowcase.view.adapter.PeopleAdapter;
import com.urban.mvvmshowcase.view.vmwrapper.AndroidPersonListViewModel;
import com.urban.mvvmshowcase.viewmodel.PersonListViewModel;

public class MainActivity extends AppCompatActivity implements PersonListViewModel.PeopleListObserver {
    private AndroidPersonListViewModel mPersonViewModel;
    private PeopleAdapter mPeopleAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ActivityMainBinding viewBinding = DataBindingUtil.setContentView(this, R.layout.activity_main);
        configureViewModel(viewBinding);
        configureListAdapter(viewBinding);
    }

    private void configureListAdapter(ActivityMainBinding viewBinding) {
        mPeopleAdapter = new PeopleAdapter();
        viewBinding.peopleList.setLayoutManager(new LinearLayoutManager(this));
        viewBinding.peopleList.setAdapter(mPeopleAdapter);
    }

    private void configureViewModel(ActivityMainBinding viewBinding) {
        mPersonViewModel = (AndroidPersonListViewModel) getLastCustomNonConfigurationInstance();
        if (mPersonViewModel == null) {
            mPersonViewModel = new AndroidPersonListViewModel(FakePersonRepository.create());
        }
        mPersonViewModel.setListObserver(this);
        viewBinding.setVm(mPersonViewModel);
    }

    @Override
    public Object onRetainCustomNonConfigurationInstance() {
        return mPersonViewModel;
    }

    @Override
    protected void onResume() {
        super.onResume();
        mPersonViewModel.onShow();
    }

    @Override
    protected void onPause() {
        super.onPause();
        mPersonViewModel.onHide();
    }

    @Override
    public void onPeopleListChanged() {
        mPeopleAdapter.setPeople(mPersonViewModel.getPeople());
    }
}

File: app/src/main/java/com/urban/mvvmshowcase/person/list/MainActivity.kt
package com.urban.mvvmshowcase.person.list

import android.databinding.DataBindingUtil
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import com.urban.mvvmshowcase.R
import com.urban.mvvmshowcase.android.MvvmApplication
import com.urban.mvvmshowcase.databinding.ActivityMainBinding
import com.urban.mvvmshowcase.person.Person
import com.urban.mvvmshowcase.person.create.CreatePersonActivity
import com.urban.mvvmshowcase.android.base.AbstractViewModelActivity
import com.urban.mvvmshowcase.person.list.adapter.ClickListener
import com.urban.mvvmshowcase.person.list.adapter.PeopleAdapter

class MainActivity : AbstractViewModelActivity<PersonListViewModel>(),
        ClickListener, PersonListViewAccess {
    lateinit var listAdapter: PeopleAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val viewBinding = DataBindingUtil.setContentView<ActivityMainBinding>(this, R.layout.activity_main)
        listAdapter = PeopleAdapter(this, vm().people)

        viewBinding.vm = vm()
        viewBinding.view = this
    }

    override fun onResume() {
        super.onResume()
        vm().viewAccess = this
        vm().loadPeople()
    }

    override fun onPause() {
        super.onPause()
        vm().viewAccess = null
        vm().cancelLoading()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_list, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.item_save -> CreatePersonActivity.start(this).let { true }
            else -> super.onOptionsItemSelected(item)
        }
    }

    override fun notifyDataChanged() = listAdapter.notifyDataSetChanged()

    override fun onPersonClick(person: Person) = CreatePersonActivity.startEdit(this, person)

    override fun createViewModel(): PersonListViewModel =
            PersonListViewModel(MvvmApplication.get(this).personRepository, this)
}